hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/scheduler/capacity/ResourceCalculationDriver.java

```diff
@@ -0,0 +1,368 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hadoop.yarn.server.resourcemanager.scheduler.capacity;
+
+import org.apache.commons.lang3.tuple.ImmutablePair;
+import org.apache.commons.lang3.tuple.Pair;
+import org.apache.hadoop.thirdparty.com.google.common.collect.ImmutableSet;
+import org.apache.hadoop.yarn.server.resourcemanager.scheduler.capacity.QueueCapacityVector.QueueCapacityVectorEntry;
+import org.apache.hadoop.yarn.server.resourcemanager.scheduler.capacity.QueueCapacityVector.ResourceUnitCapacityType;
+
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+
+import static org.apache.hadoop.yarn.api.records.ResourceInformation.MEMORY_URI;
+
+/**
+ * Drives the main logic of resource calculation for all children under a parent queue. Acts as a
+ * bookkeeper of disposable update information that is used by all children under a common parent.
+ */
+public class ResourceCalculationDriver {
+  protected static final Set<ResourceUnitCapacityType> CALCULATOR_PRECEDENCE =
+      ImmutableSet.of(
+          ResourceUnitCapacityType.ABSOLUTE,
+          ResourceUnitCapacityType.PERCENTAGE,
+          ResourceUnitCapacityType.WEIGHT);
+
+  protected final QueueResourceRoundingStrategy roundingStrategy =
+      new DefaultQueueResourceRoundingStrategy(CALCULATOR_PRECEDENCE);
+  protected final CSQueue parent;
+  protected final QueueCapacityUpdateContext updateContext;
+  protected final Map<ResourceUnitCapacityType, AbstractQueueCapacityCalculator> calculators;
+  protected final Collection<String> definedResources;
+
+  protected final Map<String, ResourceVector> overallRemainingResource = new HashMap<>();
+  protected final Map<String, ResourceVector> batchRemainingResource = new HashMap<>();
+  // Used by ABSOLUTE capacity types
+  protected final Map<String, ResourceVector> normalizedResourceRatio = new HashMap<>();
+  // Used by WEIGHT capacity typet js
+  protected final Map<String, Map<String, Float>> sumWeightsPerLabel = new HashMap<>();
+
+  protected String currentResourceName;
+  protected AbstractQueueCapacityCalculator currentCalculator;
+  protected CSQueue currentChild;
```

Comment:
I don't think these should be fields and internal states, it is much easier to get out of sync with the callers in the case of complex recursion. I think these should be arguments for the apropriate methods, if you are concerened about the number of arguments these can be wrapped into an object. But this way you split the traversal responsibility between the components. The component which makes the actual traversal and recursion has a state about where we are in the tree (resource name, child etc), and it is set onto this driver, which will behave differently based on these settings, so this object also tracks it. The problem if any of these fields is not cleared after a step in the recurison, the actual traversal and the state of this oject will be out of sync, and it will be a nightmare to debug it.

Reply:
I agree, replaced it with a context object that is passed as argument.